server/player/player.go

```diff
@@ -557,11 +557,39 @@ func (p *Player) Hurt(dmg float64, source damage.Source) (float64, bool) {
 		p.Exhaust(0.1)
 
 		damageToArmour := int(math.Max(math.Floor(dmg/4), 1))
+		var damageToAttacker int
+		thornsArmour := map[int]item.Stack{}
 		for slot, it := range p.armour.Slots() {
 			if _, ok := it.Item().(item.Durable); ok {
+				if e, ok := it.Enchantment(enchantment.Thorns{}); ok && rand.Float64() < float64(e.Level())*0.15 {
+					damageToArmour++
+					thornsArmour[slot] = it
+					if e.Level() > 10 {
+						damageToAttacker += e.Level() - 10
+					} else {
+						damageToAttacker += 1 + rand.Intn(4)
+					}
+				}
 				_ = p.armour.Inventory().SetItem(slot, p.damageItem(it, damageToArmour))
 			}
 		}
+		// Damage a random thorns piece that the user is wearing.
+		for slot, item := range thornsArmour {
+			_ = p.armour.Inventory().SetItem(slot, p.damageItem(item, 2))
+
+			if damageToAttacker > 0 {
+				var attacker world.Entity
+				if s, ok := source.(damage.SourceEntityAttack); ok {
+					attacker = s.Attacker
+				} else if s, ok := source.(damage.SourceProjectile); ok {
+					attacker = s.Owner
+				}
+				if l, ok := attacker.(entity.Living); ok {
+					l.Hurt(float64(damageToAttacker), damage.SourceThorns{Owner: attacker})
+				}
+			}
+			break
+		}
```

Comment:
```suggestion
		for slot, item := range thornsArmour {
			_ = p.armour.Inventory().SetItem(slot, p.damageItem(item, 2))

			if damageToAttacker > 0 {
				var attacker world.Entity
				if s, ok := source.(damage.SourceEntityAttack); ok {
					attacker = s.Attacker
				} else if s, ok := source.(damage.SourceProjectile); ok {
					attacker = s.Owner
				}
				if l, ok := attacker.(entity.Living); ok {
					l.Hurt(float64(damageToAttacker), damage.SourceThorns{Owner: attacker})
				}
			}
			break
		}
```

This guarantees the same item will always take the damage penalty